packages/network-controller/src/NetworkController.ts

```diff
@@ -991,201 +1293,627 @@ export class NetworkController extends BaseController<
+   */
+  getNetworkConfigurationByChainId(
+    chainId: Hex,
+  ): NetworkConfiguration | undefined {
+    return this.state.networkConfigurationsByChainId[chainId];
+  }
+
+  /**
+   * Returns the network configuration that contains an RPC endpoint with the
+   * given network client ID.
+   *
+   * @param networkClientId - The network client ID to use as a key.
+   * @returns The network configuration if one exists, or undefined.
    */
   getNetworkConfigurationByNetworkClientId(
     networkClientId: NetworkClientId,
   ): NetworkConfiguration | undefined {
-    if (isInfuraNetworkType(networkClientId)) {
-      const rpcUrl = `https://${networkClientId}.infura.io/v3/${
-        this.#infuraProjectId
-      }`;
-      return {
-        rpcUrl,
-        ...BUILT_IN_NETWORKS[networkClientId],
-      };
-    }
-
-    return this.state.networkConfigurations[networkClientId];
+    return this.#networkConfigurationsByNetworkClientId.get(networkClientId);
   }
 
   /**
-   * Adds a new custom network or updates the information for an existing
-   * network.
+   * Creates and registers network clients for the collection of Infura and
+   * custom RPC endpoints that can be used to make requests for a particular
+   * chain, storing the given configuration object in state for later reference.
    *
-   * This may involve updating the `networkConfigurations` property in
-   * state as well and/or adding a new network client to the network client
-   * registry. The `rpcUrl` and `chainId` of the given object are used to
-   * determine which action to take:
-   *
-   * - If the `rpcUrl` corresponds to an existing network configuration
-   * (case-insensitively), then it is overwritten with the object. Furthermore,
-   * if the `chainId` is different from the existing network configuration, then
-   * the existing network client is replaced with a new one.
-   * - If the `rpcUrl` does not correspond to an existing network configuration
-   * (case-insensitively), then the object is used to add a new network
-   * configuration along with a new network client.
-   *
-   * @param networkConfiguration - The network configuration to add or update.
-   * @param options - Additional configuration options.
-   * @param options.referrer - Used to create a metrics event; the site from which the call originated, or 'metamask' for internal calls.
-   * @param options.source - Used to create a metrics event; where the event originated (i.e. from a dapp or from the network form).
-   * @param options.setActive - If true, switches to the network upon adding or updating it (default: false).
-   * @returns The ID for the added or updated network configuration.
+   * @param fields - The object that describes the new network/chain and lists
+   * the RPC endpoints which front that chain.
+   * @returns The newly added network configuration.
+   * @throws if any part of `fields` would produce invalid state.
+   * @see NetworkConfiguration
```

Comment:
nit:
```suggestion
   * @see {@link NetworkConfiguration}
```

Reply:
Fixed in [a600e92](https://github.com/MetaMask/core/pull/4286/commits/a600e9244be9f77c88e9369f6aab7c833b0374a1).